firedrake/variational_solver.py

```diff
@@ -239,6 +234,19 @@ def set_transfer_manager(self, manager):
         :raises ValueError: if called after the transfer manager is setup.
         """
         self._ctx.transfer_manager = manager
+        problem = self._problem
+        ctx = self._ctx
+        seen = set()
+        coefficients = problem.F.coefficients() + problem.J.coefficients()
+        if problem.Jp is not None:
+            coefficients = coefficients + problem.Jp.coefficients()
+        for val in chain(coefficients, (bc.function_arg for bc in problem.bcs)):
+            if isinstance(val, (function.Function, cofunction.Cofunction)):
+                V = val.function_space()
+                if V not in seen:
+                    dm = V.dm
+                    dmhooks.push_appctx(dm, ctx)
```

Comment:
@wence- we need this to avoid the warning "Creating a new transfer manager, this could be expensive". I am not sure where this should go.

Reply:
So I think what you need is for the appcontext to be set on all the function space DMs when you call `SNESSolve` (rather than just the DM corresponding to the functionspace that the solution lives in).

So I think you would do it with a `dmhooks.add_hooks` call in `NLVS.solve` (similar to how the appctx is attached to the solution's DM). That manages removing the appctx afterwards (rather than this approach which just continues to stack things up and will hold on to the ctx forever).